### PR TITLE
Clear stale results in R test explorer at start of run

### LINKED ISSUE #15019: R test explorer: results can be misleading after a run that doesn't finish cleanly
Pulling out bits from 2 kitchen sink issues to be the target of a PR:

From https://github.com/posit-dev/positron/issues/2638#issuecomment-4248479321:

If the test run just fails right out of the gate, the TEST RESULTS just shows this:

```
* The test run did not record any output. 
```

In the example I just encountered, it's because a dependency was not installed (a situation I was deliberately probing for some devtools work):

```
> test()
ℹ Testing devtools
Starting 2 test processes.
✔ | F W  S  OK | Context
⠦ [ FAIL 0 | WARN 0 | SKIP 0 | PASS 0 ] Starting up...                                       
Run `rlang::last_trace()` to see where the error occurred.
Error:
! testthat subprocess failed to start.
  Error in load_imports(path) :
  The package "ellipsis" (>= 0.3.2) is required.
  Calls: callr_startup_hook ... load_imports -> deps_check_installed -> check_installed
  Calls: callr_startup_hook ... load_imports -> deps_check_installed -> check_installed
```

We should be surfacing more actionable info in TEST RESULTS.

I also wonder if previous results should be cleared vs. leaving somewhat confusing green check marks behind. (I'm not 100% sure about this, in terms of how the test explorer feature/API is intended to be used, but worth thinking through.)

From https://github.com/posit-dev/positron/issues/2636#issuecomment-3599594467

If you introduce an error in the package, e.g. that makes load_all() fail, and you rerun some tests ... no tests are actually run, but this is not obvious to the user. I.e. it feels like tests that were passing are still passing or those that were failing are still failing. We need to surface any problem that prevents tests from running.

### PR DESCRIPTION
**Pull request by jennybc**
Closes #15019
Partial progress on #14668

This PR rationalizes what happens at the start of a test run and when a test run doesn't complete cleanly. This is not about tests passing or failing, but about the test machinery being stale or fouled up.

First, before running tests, we now mark every leaf test that's about to run as "queued", eagerly clearing any result from a previous run up front. Here's how a large test suite looks now at the start of "run all tests". Notice how everything immediately switches to this yellow "Queued" status. This is the main improved behaviour every user will notice!

Screenshot 2026-07-20 at 12 00 59 PM

If the run ends early, un-run tests show as "Skipped" instead of retaining their out-of-date pass/fail status from before. (I think "Not yet run" would be a more appropriate status than "Skipped", but this behaviour comes from the VS Code core, not in positron-r, so I'm living with it for now.) This requires doing more work eagerly (parsing, changing test status, enqueing), but we're not doing any *new* work. Just doing the same old work in a more useful order. I attach before/after screenshots of this in a comment below.

There's a new e2e test of the clear-on-run behaviour and I verified that it fails before the changes and passes after.

Second, another fix motivated by my manual experimentation. Now we explicitly look for testthat's `end_reporter` event. If the R process exits without that, we know it's an unclean exit and reveal more detail in the TEST RESULTS area. This output can be sort of verbose and ugly, but it's a vast improvement over "The test run did not record any output." and it's *actionable*.

This can happen via some edge cases (failed `load_all()` or even a segfault). This replaces a brittle "Execution halted" match on stdout with a more structured approach based on `end_reporter`. No formal test of this yet, because putting more useful info in TEST RESULTS is the next work item and that's when I will add the infrastructure to inspect it. I attach a screenshot of the new behaviour below.

#### New Features

- The R test explorer proactively clears existing test statuses at the start of a partial or full run, to avoid displaying a mix of current and stale results. Runs that finish uncleanly reveal more actionable information in the TEST RESULTS panel.

#### Bug Fixes

- N/A

### Validation Steps

Run a whole, large test suite and behold how all tests snap to "Queued" at the very start.

Break `load_all()` and run tests. Notice that no pre-existing passing/failing test statuses persist. Everything is marked as "Skipped". You can break `load_all()` by putting `stop("breaking load_all()")` somewhere like the start of `tests/testthat/setup.R` or `test/testthat/helper.R`. Notice the new info surfaced in TEST RESULTS.

Screenshot 2026-07-20 at 12 10 23 PM

If you want to play with a partial test run, drop this into a test file in an existing R package:

```
test_that("passing test before the quit", {
  expect_true(TRUE)
})

test_that("test that crashes R", {
  skip("not really interested in studying a crash")
  rlang::node_car(1L)
})

test_that("test with uncaught error", {
  stop("oops")
})

stop("stop testing in middle of file")

test_that("failing test after the quit", {
  expect_true(FALSE)
})
```

Run just this test file by clicking its "Run test" play button in the explorer. Comment / uncomment the `stop()` in order to witness what happens with the status of the `test_that()` *after* the `stop()`.

@:test-explorer @:web @:win

### CODE REVIEW

**Comment by jennybc** [comment]
The promised before vs. after re: stale test results.

Basic move: I fully run a test file to get all tests marked with their status. Then I un-comment a `stop()` in the middle of the file that leaves the last test un-run. 

With released Positron, the old test status ("Failing" in this example) is retained, causing the test results in that file to be a mix of current and stale.

With dev Positron, the old test status is set to "Queued" then "Skipped", leaving the results in a rational state.

Screenshot 2026-07-20 at 11 49 22 AM
Screenshot 2026-07-20 at 11 49 37 AM

Screenshot 2026-07-20 at 11 48 14 AM
Screenshot 2026-07-20 at 11 48 45 AM

File: extensions/positron-r/resources/testing/r.pkg.test.explorer.fixture/tests/testthat/test-early-stop.R [hunk 1]
```diff
@@ -0,0 +1,13 @@
+test_that("runs before the stop", {
+  expect_true(TRUE)
+})
+
+# The e2e test drops a "STOP" sentinel in tests/testthat/ so this top-level
+# stop() ends the file early, leaving the test below unrun.
```
Comment (jennybc) [inline extensions/positron-r/resources/testing/r.pkg.test.explorer.fixture/tests/testthat/test-early-stop.R:6]: Rigging it this way lets me have this file in the fixture we use to test the explorer, but this weird `stop()` is only activated by the presence of a sentinel file named `STOP`.

File: extensions/positron-r/src/testing/runner.ts [hunk 2]
```diff
@@ -5,16 +5,24 @@
...
+import { ItemType, TestingTools } from './util-testing';
+import { parseTestsFromFile } from './parser';
 import { LOGGER } from '../extension';
 
-function maybeEnqueue(
-	test: vscode.TestItem,
-	queue: vscode.TestItem[],
-	excludeSet: readonly vscode.TestItem[] | undefined
+// Paint the UI state of every leaf test under `item` as "queued", recursing through containers.
+function markLeafTestsQueued(
```
Comment (jennybc) [inline extensions/positron-r/src/testing/runner.ts:13]: The UI state of a container (e.g. a test file or a `describe()`) "bubbles up" passively from its children. So we mark leaves as "queued", but not the containers themselves.

File: extensions/positron-r/src/testing/runner-testthat.ts [hunk 3]
```diff
@@ -220,14 +213,31 @@ export async function runThatTest(
...
+					case 'end_reporter':
+						sawEndReporter = true;
+						break;
 				}
 			});
-		childProcess.once('exit', () => {
-			stdout += childProcess.stderr.read();
-			if (stdout.includes('Execution halted')) {
-				reject(Error(stdout));
+		childProcess.once('exit', (code, signal) => {
```
Comment (juliasilge) [inline extensions/positron-r/src/testing/runner-testthat.ts:221]: IIUC `exit` can fire before stdout has fully drained, so `sawEndReporter` may still be false even if `end_reporter` was emitted. Do you think it would make more sense to use `close` here for completion?

**Review by juliasilge — APPROVED** [review]
I've got one question about how we're computing `sawEndReporter` here, but in terms of how it's working, this is really nice set of improvements!

Image

File: extensions/positron-r/src/testing/runner-testthat.ts (hunk 3 above)
Comment (jennybc) [reply extensions/positron-r/src/testing/runner-testthat.ts:221]: It's a good thing to push on and if it were just this PR, I'd switch from `exit` to `close`. But I've already got a second PR in flight that also involves this handler and makes it possible to stop a test run. All things considered, I think using `exit` sets us up for a better user experience.

Just to record the thought process in case I need to revisit this decision:

Yes, `exit` can in theory fire before stdout is drained. That would cause a spurious "ended before completing" note on a run that was actually fine. I haven't seen this in practice and we can certainly revisit if users start to experience this.

I *think* this is unlikely, because `end_reporter` is emitted by testthat before R's normal shutdown begins, which has its own latency. So I predict that `end_reporter` will land well before exit fires almost all the time.

The next piece of work makes it possible to cancel a run, where R can spawn children (such as parallel testthat workers or a `system()` call). On the happy path, cancellation will also kill those children. But if any such process gets orphaned, they'd hold on to the stdout pipe and keep `close` from ever happening. That would cause the test run to never "settle" in the UI.
